packages/mdc-menu-surface/foundation.js

```diff
@@ -0,0 +1,557 @@
+    if (anchorWidth / surfaceWidth > numbers.ANCHOR_TO_MENU_SURFACE_WIDTH_RATIO) {
+      horizontalAlignment = 'center';
+    }
+
+    // Adjust vertical origin when menu surface is positioned with significant offset from anchor. This is done so that
+    // scale animation is "anchored" on the anchor.
+    if (!(this.anchorCorner_ & MenuSurfaceCornerBit.BOTTOM) &&
+        Math.abs(verticalOffset / surfaceHeight) > numbers.OFFSET_TO_MENU_SURFACE_HEIGHT_RATIO) {
+      const verticalOffsetPercent = Math.abs(verticalOffset / surfaceHeight) * 100;
+      const originPercent = (corner & MenuSurfaceCornerBit.BOTTOM)
+        ? 100 - verticalOffsetPercent : verticalOffsetPercent;
+      verticalAlignment = Math.round(originPercent * 100) / 100 + '%';
+    }
+
+    this.adapter_.setTransformOrigin(`${horizontalAlignment} ${verticalAlignment}`);
+    this.adapter_.setPosition(position);
+    this.adapter_.setMaxHeight(maxMenuSurfaceHeight ? maxMenuSurfaceHeight + 'px' : '');
+
+    // Clear measures after positioning is complete.
+    this.measures_ = null;
+  }
+
+  /**
+   * Open the menu surface.
+   * @param {{focusIndex: ?number}=} options
+   */
+  open({focusIndex = null} = {}) {
+    this.adapter_.saveFocus();
+
+    if (!this.quickOpen_) {
+      this.adapter_.addClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_OPEN);
+    }
+
+    this.animationRequestId_ = requestAnimationFrame(() => {
+      this.dimensions_ = this.adapter_.getInnerDimensions();
+      this.autoPosition_();
+      this.adapter_.addClass(MDCMenuSurfaceFoundation.cssClasses.OPEN);
+      this.focusOnOpen_(focusIndex);
+      this.adapter_.registerBodyClickHandler(this.documentClickHandler_);
+      if (!this.quickOpen_) {
+        this.openAnimationEndTimerId_ = setTimeout(() => {
+          this.openAnimationEndTimerId_ = 0;
+          this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_OPEN);
+        }, numbers.TRANSITION_OPEN_DURATION);
+      }
+    });
+    this.isOpen_ = true;
+  }
+
+  /**
+   * Closes the menu surface.
+   */
+  close() {
+    this.adapter_.deregisterBodyClickHandler(this.documentClickHandler_);
+
+    if (!this.quickOpen_) {
+      this.adapter_.addClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_CLOSED);
+    }
+
+    requestAnimationFrame(() => {
+      this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.OPEN);
+      if (!this.quickOpen_) {
+        this.closeAnimationEndTimerId_ = setTimeout(() => {
+          this.closeAnimationEndTimerId_ = 0;
+          this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_CLOSED);
+          this.adapter_.notifyClose();
+        }, numbers.TRANSITION_CLOSE_DURATION);
+      }
+    });
+
+    this.isOpen_ = false;
+    this.shouldRestoreFocus_();
+  }
+
+  /**
+   * If the user is focused on or within the menu surface when it is closed, the last focused
+   * element when the menu surface was opened is focused.
```

Comment:
The last focused element when the menu surface was opened should regain focus, if the user is focused on or within the menu surface when it is closed.